### ISSUE TITLE
JSON schema only validates abstract attributes for submodel elements

### ISSUE THREAD

**Issue by jkhsjdhjs** [admin-shell-io/aas-specs#251]
The current JSON schema only validates the abstract SubmodelElement attributes like `idShort` and `semanticId`. This is because wherever one or multiple SubmodelElements are expected, the abstract SubmodelElement definition is referenced.

Take for example the Submodel object and its `submodelElements` property. It is defined as an array of `#/definitions/SubmodelElement`.

```json
"submodelElements": {
    "type": "array",
    "items": {
        "$ref": "#/definitions/SubmodelElement"
    },
    "minItems": 1
}
```

However, `#/definitions/SubmodelElement` is simply the definition of all abstract attributes each Submodel has. It should **only** be referenced by the individual SubmodelElements, such as Operation.

```json
"SubmodelElement": {
    "allOf": [
        {
            "$ref": "#/definitions/Referable"
        },
        {
            "$ref": "#/definitions/HasKind"
        },
        {
            "$ref": "#/definitions/HasSemantics"
        },
        {
            "$ref": "#/definitions/Qualifiable"
        },
        {
            "$ref": "#/definitions/HasDataSpecification"
        }
    ],
}
```

The XML schema doesn't has this issue, since it has a group `SubmodelElement_choice`, which can be any of the possible Submodel Elements. Such a group should be introduced in the JSON schema as well.
```xml
<xs:group name="submodelElement_choice">
    <xs:choice>
        <xs:element name="relationshipElement" type="relationshipElement_t"/>
        <xs:element name="annotatedRelationshipElement" type="annotatedRelationshipElement_t"/>
        <xs:element name="basicEventElement" type="basicEventElement_t"/>
        <xs:element name="blob" type="blob_t"/>
        <xs:element name="capability" type="capability_t"/>
        <xs:element name="entity" type="entity_t"/>
        <xs:element name="file" type="file_t"/>
        <xs:element name="multiLanguageProperty" type="multiLanguageProperty_t"/>
        <xs:element name="operation" type="operation_t"/>
        <xs:element name="property" type="property_t"/>
        <xs:element name="range" type="range_t"/>
        <xs:element name="referenceElement" type="referenceElement_t"/>
        <xs:element name="submodelElementCollection" type="submodelElementCollection_t"/>
        <xs:element name="submodelElementList" type="submodelElementList_t"/>
    </xs:choice>
</xs:group>
```

The issue described above is also present for the `DataElement` definition, possibly for others as well. `SubmodelElement` and `DataElement` are the only definitions I checked.

**Comment by mristin** [comment]
@jkhsjdhjs that is the constraint we initially also introduced during the development, but took it out eventually as it produced ugly/unusable code when used together with OpenAPI specs.

@sebbader-sap could you please have a closer look? Perhaps it would be worthwhile to have two schema generators, one for JSON Schema and another one for OpenAPI?

**Comment by jkhsjdhjs** [comment]
Ah, I see. Yeah it would be good to have separate generators then, because otherwise submodel-specific attributes are just not validated, such as `lastUpdate`, `minInterval` and `maxInterval` against their pattern for example.

I also recommend setting `additionalProperties` to `false` for each definition of the JSON schema, as it allows for stricter validation. I'm not creating a new issue for this just now, as `additionalProperties: false` would in the current state break the validation of every SubmodelElement. I'll await your decision here first.

**Comment by mristin** [comment]
@jkhsjdhjs we also had `additionalProperties: false` 😀. The problem is that many users introduce their own extensions, and for simplicity attach them as JSON properties.

**Comment by jkhsjdhjs** [comment]
Hmm, should these custom extensions pass schema validation though? I could imagine that some JSON parsers go through all properties of an object and fail on unknown properties. These JSON parsers wouldn't have a problem handling non-extended schema-compliant data. To me it would make sense if users who use custom extensions would extend the schema for validation of their extensions.

**Comment by mristin** [comment]
@jkhsjdhjs 

> To me it would make sense if users who use custom extensions would extend the schema for validation of their extensions.

I agree.

@BirgitBoss please consider this a finding to be discussed for the next version of the schema.

**Comment by BirgitBoss** [comment]
Thanks for the input, we will discuss it in the Working Stream AAS of the IDTA

**Comment by sebbader-sap** [comment]
@mristin we are struggling with OpenApi 3.0.3 anyway, and think about upgrading to 3.1.0. This version shall be 100% JSON Schema compliant, at least that is what I read. 
Keeping this in mind, it might not be necessary to create different OpenApi and Json Schema generators. Nothing has been decided yet, though.

**Comment by sebbader-sap** [comment]
Coming back to this issue, I feel that hoping for OpenAPI v3.1.0 actually does not help us here. As you, @jkhsjdhjs , stated above, the `allOf` in `SubmodelElements` invalidates any kind of SubmodelElement-specific attribute validation.
The following examples show the problem:
1. String where String is expected:
```
{ 
   "modelType":"Property",
   ...
   "value": "some value"
}
```
Expected behaviour: passes validation
Observed behaviour: passes validation (this is the good case)

2. Object where String is expected:
```
{ 
   "modelType":"Property",
   ...
   "value": { "key": "some value"}
}
```
Expected behaviour: violation detected
Observed behaviour: passes validation (not good)

3. Array where Array is expected:
```
{ 
   "modelType":"SubmodelElementCollection",
   ...
   "value": [ { "modelType": "Property", ...} ]
}
```
Expected behaviour: passes validation
Observed behaviour: passes validation (good)

4. Object where Array is expected:
```
{ 
   "modelType":"SubmodelElementCollection",
   ...
   "value": { "modelType": "Property", ...} 
}
```
Expected behaviour: violation detected
Observed behaviour: passes validation (not good)

Explanation: The validator only checks for SubmodelElements only agains the attributes of Referable, HasKind, HasSemantics, HasQualifiable, and HasDataSpecification but *not* against oneOf Property/SubmodelElementCollection/...

**Comment by sebbader-sap** [comment]
I tried a pattern with `oneOf` for SubmodelElements (renaming them to `AbstractSubmodelElements`:
```
    ...
    "AbstractSubmodelElement": {
      "allOf": [
        {
          "$ref": "#/definitions/Referable"
        },
        {
          "$ref": "#/definitions/HasKind"
        },
        {
          "$ref": "#/definitions/HasSemantics"
        },
        {
          "$ref": "#/definitions/Qualifiable"
        },
        {
          "$ref": "#/definitions/HasDataSpecification"
        }
      ]
    },
    "SubmodelElement": {
      "oneOf": [
        {
          "$ref": "#/definitions/Property"
        },
        ...
        {
          "$ref": "#/definitions/SubmodelElementCollection"
        }
      ],
    ...
    "Property": {
      "allOf": [
        {
          "$ref": "#/definitions/AbstractSubmodelElement"
        },
        {
          "properties": {
            ....
            "value": { ... }
       }
    ]
  },
    ...
    "SubmodelElementCollection": {
      "allOf": [
        {
          "$ref": "#/definitions/AbstractSubmodelElement"
        },
        {
          "properties": {
            ....
            "value": { ... }
       }
    ]
  }
```

The problem with this pattern is not that it would be ugly (which I would accept) but that it creates cycles as e.g. SubmodelElementCollection isA SubmodelElement but also has one (or more) SubmodelElement in its value position which could be a SubmodelElementCollection which has one (or more) SubmodelElement in its value position which ... and so on.
Our JSON Schema Validator (more precisely our OpenAPI validator) reacts with an exception. Which I think is the correct behaviour, and that the problem - in form of the cycles - is on the schema/metamodel part.

**Comment by sebbader-sap** [comment]
Which brings me currently to the point that we have a severe problem here which we need to solve before the final V3.0...

**Comment by mristin** [comment]
@sebbader-sap on a side note, I think you have a subtle bug here, please double-check:
> The problem with this pattern is not that it would be ugly (which I would accept) but that it creates cycles as e.g. SubmodelElementCollection isA SubmodelElement but also has one (or more) SubmodelElement in its value position which could be a SubmodelElementCollection which has one (or more) SubmodelElement in its value position which ... and so on.

`SubmodelElementCollection` inherits from `SubmodelElement`, but its value is an array of `AbstractSubmodelElement`. This is still a cycle, though.

This worked with C# and Python JSON schema validators (IIRC, it's been a while), but was not properly implemented in OpenAPI-related tools.

**Comment by sebbader-sap** [comment]
> This worked with C# and Python JSON schema validators (IIRC, it's been a while),

I am not aware how the validator implementations cover cycles (recursions?), only know that our one doesn't like it. Probably more clever ones can manage them... Given that this is the case, would you say that my proposal above is actually a solution to the original problem?
And if so, can we change the generated JSON Schema structures to this pattern?

**Comment by mristin** [comment]
@sebbader-sap 
> I am not aware how other validator implementations cover cycles (recursions?), [...]

I suppose they implement some form of forward declaration akin to what C++ and other compilers do.

> Given that this is the case, would you say that my proposal above is actually a solution to the original problem?

Yes, your proposal should work for JSON Schema. Maybe we can structure it in such a way that you can merely remove lines to obtain an OpenAPI schema?

@jkhsjdhjs could I kindly ask you to test @sebbader-sap solution on a toy example? I won't be in the office till Feb 18th, but it would be great if we could move forward with this issue in the meanwhile.

**Comment by s-heppner** [comment]
This has been fixed with V3.0, so I'm closing this issue.
